instrumentation/opentelemetry-instrumentation-sqlalchemy/tests/test_sqlalchemy_metrics.py

```diff
@@ -0,0 +1,254 @@
+            key=lambda m: m.name,
+        )
+
+    def assert_metric_expected(
+        self,
+        metric: Metric,
+        expected_data_points: Sequence[DataPointT],
+        est_value_delta: Optional[float] = 0,
+    ):
+        self.assertEqual(
+            len(expected_data_points), len(metric.data.data_points)
+        )
+        for expected_data_point in expected_data_points:
+            self.assert_data_point_expected(
+                expected_data_point, metric.data.data_points, est_value_delta
+            )
+
+    # pylint: disable=unidiomatic-typecheck
+    @staticmethod
+    def is_data_points_equal(
+        expected_data_point: DataPointT,
+        data_point: DataPointT,
+        est_value_delta: Optional[float] = 0,
+    ):
+        if type(expected_data_point) != type(data_point) or not isinstance(
+            expected_data_point, (HistogramDataPoint, NumberDataPoint)
+        ):
+            return False
+
+        values_diff = None
+        if isinstance(data_point, HistogramDataPoint):
+            values_diff = abs(expected_data_point.sum - data_point.sum)
+        elif isinstance(data_point, NumberDataPoint):
+            values_diff = abs(expected_data_point.value - data_point.value)
+
+        return (
+            values_diff <= est_value_delta
+            and expected_data_point.attributes == dict(data_point.attributes)
+        )
+
+    def assert_data_point_expected(
+        self,
+        expected_data_point: DataPointT,
+        data_points: Sequence[DataPointT],
+        est_value_delta: Optional[float] = 0,
+    ):
+        is_data_point_exist = False
+        for data_point in data_points:
+            if self.is_data_points_equal(
+                expected_data_point, data_point, est_value_delta
+            ):
+                is_data_point_exist = True
+                break
+
+        self.assertTrue(
+            is_data_point_exist,
+            msg=f"Data point {expected_data_point} does not exist",
+        )
+
+    @staticmethod
+    def create_number_data_point(value, attributes):
+        return NumberDataPoint(
+            value=value,
+            attributes=attributes,
+            start_time_unix_nano=0,
+            time_unix_nano=0,
+        )
+
+    def assert_pool_idle_used_expected(self, pool_name, idle, used):
+        metrics = self.get_sorted_metrics()
+        self.assertEqual(len(metrics), 1)
+        self.assert_metric_expected(
+            metrics[0],
+            [
+                self.create_number_data_point(
+                    value=idle,
+                    attributes={"pool.name": pool_name, "state": "idle"},
+                ),
+                self.create_number_data_point(
+                    value=used,
+                    attributes={"pool.name": pool_name, "state": "used"},
+                ),
+            ],
+        )
+
+    def test_metrics_one_connection(self):
+        pool_name = "pool_test_name"
+        engine = sqlalchemy.create_engine(
```

Comment:
consider moving the basic engine version to test setUp

Reply:
I thought about this, but the create_engine params are not the same for all of the tests

Reply:
I'm in favor of having the basic version as setUp and test-specific changes per test but this is not a blocker by any means